runtime/src/main/cpp/Memory.cpp

```diff
@@ -2262,56 +2287,67 @@ void freezeAcyclic(ContainerHeader* rootContainer, ContainerHeaderSet* newlyFroz
   }
 }
 
-void freezeCyclic(ContainerHeader* rootContainer,
+void freezeCyclic(ObjHeader* root,
                   const KStdVector<ContainerHeader*>& order,
                   ContainerHeaderSet* newlyFrozen) {
   KStdUnorderedMap<ContainerHeader*, KStdVector<ContainerHeader*>> reversedEdges;
-  KStdDeque<ContainerHeader*> queue;
-  queue.push_back(rootContainer);
+  KStdDeque<ObjHeader*> queue;
+  queue.push_back(root);
   while (!queue.empty()) {
-    ContainerHeader* current = queue.front();
+    ObjHeader* current = queue.front();
     queue.pop_front();
-    current->unMark();
-    reversedEdges.emplace(current, KStdVector<ContainerHeader*>(0));
-    traverseContainerReferredObjects(current, [current, &queue, &reversedEdges](ObjHeader* obj) {
+    ContainerHeader* currentContainer = current->container();
+    currentContainer->unMark();
+    reversedEdges.emplace(currentContainer, KStdVector<ContainerHeader*>(0));
+    traverseContainerReferredObjects(currentContainer, [current, currentContainer, &queue, &reversedEdges](ObjHeader* obj) {
           ContainerHeader* objContainer = obj->container();
           if (canFreeze(objContainer)) {
             if (objContainer->marked())
-              queue.push_back(objContainer);
-            reversedEdges.emplace(objContainer, KStdVector<ContainerHeader*>(0)).first->second.push_back(current);
+              queue.push_back(obj);
+            // We ignore FreezableAtomicsReference during condensation, to avoid KT-33824.
+            if (!isFreezableAtomic(current) && !isFreezableAtomic(obj))
```

Comment:
`!isFreezableAtomic(obj)` is likely not required here.
Removing it will make this part more consistent with `depthFirstTraversal`.

Reply:
It just makes transposed graph explicit in fact that we shalln't follow edge both from and to freezable atomic, and if we'll do an assert will trigger.

Reply:
>  we shalln't follow edge ... to freezable atomic

This is redundant. It is enough to properly ignore edges _from_ freezable atomics in the entire condensation algorithm.

> and if we'll do an assert will trigger.

I don't think so.

Reply:
I meant `RuntimeAssert(it != reversedEdges->end(), "unknown node during condensation building");`.

Reply:
I'm still not sure that removing the discussed check will trigger this assertion. Can you provide an example?